### ISSUE TITLE
Scraper Not checking date

### ISSUE THREAD

**Issue by LindsayL** [edegraff/FundTracker#11]
We should have the scraper check the date it is scraping.

For example, today is March 8th, the CIBC website only has the current prices as of March 6th.  Yesterday it also only had the current prices as of march 6th.  This results in our scraper saving the values for March 6th incorrectly twice.  It is storing the values of march 6th for the dates 7th and 8th.  This is incorrect.

Proposal:
We should check the date and ensure we don't have an existing entry for the date before storing.  
The 3 sections that contain price-type funds each indicated the date of the current price, I'm not sure if they are in sync so we might want to check each group individually.

Extra:
I'm not sure if it would work like this... but given the situation with 6th, 7th, and 8th that currently exists, I wonder if on the 9th they will update the current price to be the 9th and, since it also shows the previous price, I wonder if it will show the previous price as of the 8th... (Aka we might be able to extract both the current price and previous to fill in the 9th and 8th respectively.)

**Comment by edegraff** [comment]
Should we only have entries for the days they update then?  I imagine the previous price will refer to the last time they updated, not the previous day.

**Comment by LindsayL** [comment]
"Should we only have entries for the days they update then?"
I think so.  It would be inaccurate to do anything else.

"I imagine the previous price will refer to the last time they updated, not the previous day."
I think you are most likely right.  In which case we wouldn't need to consider getting data from the previous price column.  Which would be best. :D  (But I guess we'll see for sure on the next update)

**Comment by LindsayL** [comment]
To confirm, the previous price does refer to the last time they were updated.

I created a function, GetValueByDate(Date d), that retrieves the value for that day based on our database only storing values on the days they update.

**Comment by edegraff** [comment]
As we discussed on Monday, we'll just have the scraper update the price every day regardless of whether it changed on CIBC.

In other news, I messed up deploying the scraper to Azure, so it did not properly store data for the last week :p  Should be fixed now.

**Comment by LindsayL** [comment]
The scraper isn't working quite right yet, sorry. xD  It's recording the wrong date/value combos.  Eg:

Fund 489, CIBC Canadian Short-Term Bond Index Fund 

On CIBC:
Current Price as of March 19: 10.1659
Previous Price as of March 18: 10.1643

In our DB:
18th: 10.1478
19th: 10.1643

So it recorded the March 18th value of 10.1643 for March 19th.  I think what's happening is we are recording the new values very early in the day (00:xx) (I assume...) and CIBC isn't actually updating it's values until later in the day, say 10:00, or maybe not even until the next day.

So I think the scraper should still run really early in the morning and record whatever it can, (because there is also the chance that CIBC skips posting values for the current day), and then it should continually re-run and check if CIBC actually changed the date for the current price, and if it did overwrite any entries that should be changed.  

A worst case scenario could look like this I think:

CIBC says current value as of Day 1, previous price as of Day 0
Day 1: current = 10.0000 previous = 7.0000

Day 2: current = 10.0000 previous = 7.0000

Day 3: current = 10.0000 previous = 7.0000

Day 4: current = 10.0000 previous = 7.0000
(So at this point our DB has all 4 days with value 10.0000)

Then CIBC updates sometime during Day 4... 
current value as of Day 3, previous price as of Day 2
Day 4: current = 15.0000 previous = 12.0000

This means that we should go change all entries for Day 3+ to 15.0000 and Days [2-3) to the previous price, 12.0000.

But I think it is quite unlikely that they would skip 2 updates, I think....  So we could maybe ignore the previous price part.  (But if we want to be really thorough we shouldn't.

This situation is part of the reason I was thinking that we should only store values for the days that CIBC updates it's current price Date.  That way the scraper would just have to check if the date changed since its previous recording and if it has do an update.  (No replacing neccesary, and with only that information the data would still look the same to anything outside of the DB due to the way the controller is currently written.

Long story short, I think either method would be a fine fix (or any other suggestions too!), I'm just not sure which would be the easiest.  (Also, for thoroughness, I wonder if we should check the date above each section on the funds page, (income funds, growth funds, managed portfolios), because I'm not sure that the current price Date would be in sync across all 3.  

(If you want me to work on any of this please let me know, I don't want to seem like I am just nit-picking and not doing anything about it, but I also didn't want to just start doing anything there since its kind of your section? xD)

**Comment by edegraff** [comment]
Ah, good catch.

How about this:
The scraper will compare the current value of a fund with the value that it's scraped, and will only add a new entry if the value has changed.  This way, there won't be entries for dates that don't have data posted to CIBC, like you want.

**Comment by LindsayL** [comment]
Sounds good to me!  

(When it finds that the value changes, can you also scrape the date that cibc says is the current date and store that date with the new values (regardless of what day it actually is)?  That way, even if CIBC is a more than 24hrs late updating, our data will still match theirs.)

**Comment by edegraff** [comment]
Sure, the scraper will check for the date now too.  The database should be correct now for the 17th, 18th, and 19th, and the updated scraper is running on Azure.

Close with 26fb22414b31abbf88da71012e6acbc935e644b8
